# Data Preprocessing

Suppose you are assigned to develop a machine learning model to predict whether an individual earns more than USD 50,000 or less in a year using the 1994 US Census Data. The datasets are sourced from the UCI Machine Learning Repository at http://archive.ics.uci.edu/ml/datasets/Census+Income.

The repository provides 5 datasets. However, each dataset is raw and does not come in the form of ABT (Analytic Base Table). The datasets are apparently not ready for predictive modeling.

The objective of this notebook is to guide you through the data preprocessing steps on the raw datasets in a sequence of exercises. The expected outcome is "clean" data that can be directly fed into any machine learning algorithm within the Scikit-Learn Python module. The clean data should look like the dataset used in this case study on our website.

# Exercises

**Exercise 0**

Read the training and test datasets directly from the data URL's. Also, since the datasets do not contain the feature names, explicitly specify them while loading in the datasets. Once you read in adultData and adultTest datasets, concatenate them into a single dataset called df.

In [1]:
import pandas as pd
import numpy as np

In [2]:
#column names details are given in website 
col_names = ['age','workclass','fnlwgt','education','education-num','marital-status','occupation','relationship','race','sex','capital-gain','capital-loss','hours-per-week','native_country','income']

In [3]:
train_df = pd.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",sep = ',', names=col_names)
train_df.head(2)

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   

        marital-status        occupation    relationship    race    sex  \
0        Never-married      Adm-clerical   Not-in-family   White   Male   
1   Married-civ-spouse   Exec-managerial         Husband   White   Male   

   capital-gain  capital-loss  hours-per-week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K

In [4]:
test_df = pd.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test",
                      sep = ',', names= col_names, skiprows=1)
test_df.head(3)

age   workclass  fnlwgt    education  education-num       marital-status  \
0   25     Private  226802         11th              7        Never-married   
1   38     Private   89814      HS-grad              9   Married-civ-spouse   
2   28   Local-gov  336951   Assoc-acdm             12   Married-civ-spouse   

           occupation relationship    race    sex  capital-gain  capital-loss  \
0   Machine-op-inspct    Own-child   Black   Male             0             0   
1     Farming-fishing      Husband   White   Male             0             0   
2     Protective-serv      Husband   White   Male             0             0   

   hours-per-week  native_country   income  
0              40   United-States   <=50K.  
1              50   United-States   <=50K.  
2              40   United-States    >50K.

In [5]:
df = pd.concat([train_df,test_df])
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

 


**--------------------------------------------------------------------------------**

**`Another way to do this:`**

url = (
    "http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",    
    "http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test",
)  # pay attention that this defines a tuple that cannot be modified


`adultData` = pd.read_csv(url[0], sep = ',', names = col_names, header = None)

`adultTest` = pd.read_csv(url[1] , sep = ',', names = col_names, skiprows = 1)

`df` = pd.concat([adultData,adultTest])

df.head()

**--------------------------------------------------------------------------------**

**Exercise 1**

Make sure the feature types match the descriptions outlined in the Data Description section. For example, confirm `age` is a numeric feature.

In [6]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native_country    object
income            object
dtype: object

**Exercise 2**

Calculate the number of missing values for each feature. Does the result surprise you?

In [7]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native_country    0
income            0
dtype: int64

**Exercise 3**

In Exercise 2, you should see zero missing value for each value. This indicates some features are coded with different labels such as "?" and "99999", instead of NaN. To provide a better overview, generate summary statistics of df. Hint: Use the describe() method with `include=np.number` and `include=np.object`. Make sure you have Python 3.6+ for this to work!

In [8]:
# Only looking in numeric columns
df.describe(include=np.number).round(2)

age      fnlwgt  education-num  capital-gain  capital-loss  \
count  48842.00    48842.00       48842.00      48842.00       48842.0   
mean      38.64   189664.13          10.08       1079.07          87.5   
std       13.71   105604.03           2.57       7452.02         403.0   
min       17.00    12285.00           1.00          0.00           0.0   
25%       28.00   117550.50           9.00          0.00           0.0   
50%       37.00   178144.50          10.00          0.00           0.0   
75%       48.00   237642.00          12.00          0.00           0.0   
max       90.00  1490400.00          16.00      99999.00        4356.0   

       hours-per-week  
count        48842.00  
mean            40.42  
std             12.39  
min              1.00  
25%             40.00  
50%             40.00  
75%             45.00  
max             99.00

In [9]:
# Only looking in string value columns
df.describe(include=np.object)

<ipython-input-9-984795f0d1fa>:2: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  df.describe(include=np.object)


workclass education       marital-status       occupation relationship  \
count      48842     48842                48842            48842        48842   
unique         9        16                    7               15            6   
top      Private   HS-grad   Married-civ-spouse   Prof-specialty      Husband   
freq       33906     15784                22379             6172        19716   

          race    sex  native_country  income  
count    48842  48842           48842   48842  
unique       5      2              42       4  
top      White   Male   United-States   <=50K  
freq     41762  32650           43832   24720

**Exercise 4**

In Exercise 3, you can see the target feature `income` has four unique values. This contradicts the definition of `income` as it should have two only labels: "<=50K" and ">50K". In this exercise, return the unique values of `income`.

In [10]:
df['income'].unique()

array([' <=50K', ' >50K', ' <=50K.', ' >50K.'], dtype=object)

**Exercise 5**

In Exercise 4, you should see income consists of 4 unique values. The values are `[' <=50K', ' >50K', ' <=50K.', ' >50K.']`. The value contains excessive white space. In this exercise:

Remove the excessive white space of income in df.
Correct the lable of income in df. In particular, relabel `>50K. and <=50K`. to `>50K and <=50K` respectively by removing .

In [11]:
df['income'] = df['income'].str.strip()
df['income'] = df['income'].str.replace('.','')


<ipython-input-11-623335e32634>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['income'] = df['income'].str.replace('.','')


In [12]:
df['income'].unique()

array(['<=50K', '>50K'], dtype=object)

**Exercise 6**

In Exercise 5, you can see that the raw (or the pre-cleaned) income column contained excessive white spaces. Can other categorical features ('workclass', 'education', 'marital-status', 'relationship', 'occupation', 'race', 'sex','native-country') have the same problem? Check which features have excessive white spaces. Remove the white spaces if necessary.

In [13]:
Categorical_col = ['workclass','education','marital-status','occupation','relationship','race','sex','native_country']
for col in Categorical_col:
    print('Unique values of "'+col+'" before removing extra space')
    print(df[col].unique(),'\n')
    
    df[col]=df[col].str.strip()
    
    print('\n','Unique values of "'+col+'" after removing extra space::',)
    print(df[col].unique(),'\n')

Unique values of "workclass" before removing extra space
[' State-gov' ' Self-emp-not-inc' ' Private' ' Federal-gov' ' Local-gov'
 ' ?' ' Self-emp-inc' ' Without-pay' ' Never-worked'] 


 Unique values of "workclass" after removing extra space::
['State-gov' 'Self-emp-not-inc' 'Private' 'Federal-gov' 'Local-gov' '?'
 'Self-emp-inc' 'Without-pay' 'Never-worked'] 

Unique values of "education" before removing extra space
[' Bachelors' ' HS-grad' ' 11th' ' Masters' ' 9th' ' Some-college'
 ' Assoc-acdm' ' Assoc-voc' ' 7th-8th' ' Doctorate' ' Prof-school'
 ' 5th-6th' ' 10th' ' 1st-4th' ' Preschool' ' 12th'] 


 Unique values of "education" after removing extra space::
['Bachelors' 'HS-grad' '11th' 'Masters' '9th' 'Some-college' 'Assoc-acdm'
 'Assoc-voc' '7th-8th' 'Doctorate' 'Prof-school' '5th-6th' '10th'
 '1st-4th' 'Preschool' '12th'] 

Unique values of "marital-status" before removing extra space
[' Never-married' ' Married-civ-spouse' ' Divorced'
 ' Married-spouse-absent' ' Separated' ' 

**Exercise 7**

The `workclass, occupation,` and `native-country` contain some missing values encoded as "?". Check the percentage of "?" in each of workclass, occupation, and native-country.

In [14]:
missingvalcol = ['workclass','occupation','native_country']
for col in missingvalcol:
    print('Column Name : '+col)
    print(df[col].value_counts(normalize=True).mul(100).round(2),'\n')

Column Name : workclass
Private             69.42
Self-emp-not-inc     7.91
Local-gov            6.42
?                    5.73
State-gov            4.06
Self-emp-inc         3.47
Federal-gov          2.93
Without-pay          0.04
Never-worked         0.02
Name: workclass, dtype: float64 

Column Name : occupation
Prof-specialty       12.64
Craft-repair         12.51
Exec-managerial      12.46
Adm-clerical         11.49
Sales                11.27
Other-service        10.08
Machine-op-inspct     6.19
?                     5.75
Transport-moving      4.82
Handlers-cleaners     4.24
Farming-fishing       3.05
Tech-support          2.96
Protective-serv       2.01
Priv-house-serv       0.50
Armed-Forces          0.03
Name: occupation, dtype: float64 

Column Name : native_country
United-States                 89.74
Mexico                         1.95
?                              1.75
Philippines                    0.60
Germany                        0.42
Puerto-Rico                    0.3

**Exercise 8**

In Exercise 7, you will notice missing values for both of the workclass and occupation features are about 5.7%. The native-country feature contains less than 2% of missing values. Note that the missing data (around more than 90%) are predominantly for the <=50K income whereas ~76% of observations pertain to the <=50K income at an aggregate level. Therefore, we shall not to impute these missing values but remove them instead as there would be minimal information loss.

In this exercise, remove the rows where workclass="?", occupation="?" and native-country="?". Check that the number of observations reduces from 48,842 to 45,222.

In [18]:
df = df[df['workclass'] != '?']
df = df[df['occupation'] != '?']
df = df[df['native_country'] != '?']

df.shape

(45222, 15)

In [20]:
df['race'].value_counts(normalize=True).mul(100).round(2)

White                 86.03
Black                  9.35
Asian-Pac-Islander     2.88
Amer-Indian-Eskimo     0.96
Other                  0.78
Name: race, dtype: float64

**Exercise 9**

In Exercise 7, notice that native-country is too granular and unbalanced. That is, close to 90% of native-country is "United-States" and the remaining 10% is made up of 40 different countries. The granularity (or the large cardinality) would yield a large number of columns when we encode native-country. You should also notice race exhibits the same problem where white accounts more than 75% of instances. In this exercise,

- For `native-country`, relabel all other countries as "Other" except "United-States".
- Likewise, relabel all other `race` as "Other" except "White".

In [24]:
df.loc[df['native_country'] != 'United-States', 'native_country'] = 'other'
df.loc[df['race'] != 'White', 'race'] = 'other'

In [29]:
print('Race',df['race'].value_counts(),'\n')
print('Native country:',df['native_country'].value_counts())

Race White    38903
other     6319
Name: race, dtype: int64 

Native country: United-States    41292
other             3930
Name: native_country, dtype: int64


**Exercise 10**

Recall that fnlwgt stands for "Final Weight" defined by the US Census. The weight is an "estimate of the number of units in the target population that the responding unit represents" This feature aims to allocate similar weights to people with similar demographic characteristics. In short, fnlwght has no predictive power.

In this exercise, remove fnlwgt from df.

In [ ]:
df = df.drop(columns='fnlwgt')
df.head(0)

**Exercise 11**

We suspect education and education-num might carry the same information. If they represent the same information, we should remove one of them (why?). To see this, run len(df['education'].unique()) and len(df['education-num'].unique()). Both should give you 16 unique values. To vindicate our suspection, make sure education and education-num indeed represent the same information. Then, drop education from df.

**Hint:** try pd.pivot_table.

In [45]:
print('Education : ',len(df['education'].unique()),'\n')
print('Education Num : ',len(df['education-num'].unique()))

Education :  16 

Education Num :  16


In [53]:
pd.pivot_table(df[['education-num', 'education']], values='education-num',  index='education')

education-num
education                  
10th                      6
11th                      7
12th                      8
1st-4th                   2
5th-6th                   3
7th-8th                   4
9th                       5
Assoc-acdm               12
Assoc-voc                11
Bachelors                13
Doctorate                16
HS-grad                   9
Masters                  14
Preschool                 1
Prof-school              15
Some-college             10

In [54]:
df = df.drop(columns='education')
df.head(0)

Empty DataFrame
Columns: [age, workclass, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native_country, income]
Index: []

#### **Remark**

In the previous exercises, we have performed heaps of data wrangling on target and categorical features. Let's focus on the continuous/numeric features: age, capital-gain, capital-loss and hours-per-week. Based on the summary statistics (in Exercise 2), age ranges 17 to 90 years old with a mean of 38.64. Therefore, we can conclude age has a reasonable range of values and hence requires no data wrangling. However, hours-per-week ranges from 1 to 99 hours with a mean of 40. We suspect 99 is used to label the "missing" instance of hours-per-week. On the other hand, it is still possible to work more than 90 hours per week. Therefore, we shall not preprocess it further. To check this, you can find the second largest value of hours-per-week is 98. Next, capital-gain ranges from 0 to 99,999 and we suspect that the missing observations are hence labeled as "99999". Is that true? We run the following code and find that a capital gain value of 99,999 always returns a higher income earner, we conjecture it might be a useful predictive value and hence we shall not remove the observations with this value.

df.loc[df['capital-gain'] == 99999.000000, 'income'].value_counts()

In [55]:
df.loc[df['capital-gain'] == 99999.000000, 'income'].value_counts()

>50K    229
Name: income, dtype: int64

**Exercise 12**

We suspect that `capital-loss =  − capital-gain` because an individual can either pay any capital gain or claim for capital loss to reduce capital gain in future (TaxBracket.org, 2017). Another possibility is to pay neither gain nor loss. Hence, it is more reasonable to record gain or loss as a single variable, rather than having them separate. Before defining such new variable, define a "mask_both" to verify that no observations were recorded as both positive capital-gain and capital-loss values. You should see there is zero count for "True" in the "mask_both".

In [57]:
mask_both = (df['capital-gain']>0) & (df['capital-loss']>0)
mask_both.value_counts()

False    45222
dtype: int64

**Exercise 13**

In light of Exercise 12, define a variable named capital which is given as `capital-gain - capital-loss`. Then remove `capital-gain` and `capital-loss` from df.

In [65]:
df['capital'] = df['capital-gain'] - df['capital-loss']
df = df.drop(columns=['capital-gain','capital-loss'])
df['capital'].describe()

count    45222.000000
mean      1012.834925
std       7530.315380
min      -4356.000000
25%          0.000000
50%          0.000000
75%          0.000000
max      99999.000000
Name: capital, dtype: float64

In [66]:
df.head(0)

Empty DataFrame
Columns: [age, workclass, education-num, marital-status, occupation, relationship, race, sex, hours-per-week, native_country, income, capital]
Index: []